test_util/cluster.py

```diff
@@ -278,6 +358,118 @@ def install_dcos(
+        # Retry on SSH command error or metric not equal to expected value.
+        retry_on_exception=(lambda exc: isinstance(exc, CalledProcessError)),
+        retry_on_result=(lambda result: not result),
+    )
+    def wait_for_mesos_metric(cluster, host, key, value):
+        """Return True when host's Mesos metric key is equal to value."""
+        return cluster.mesos_metrics_snapshot(host).get(key) == value
+
+    assert all(h.public_ip for h in cluster.hosts), (
+        'All cluster hosts must be externally reachable. hosts: {}'.formation(cluster.hosts)
+    )
+    assert cluster.bootstrap_host, 'Upgrade requires a bootstrap host'
+
+    bootstrap_url = 'http://' + cluster.bootstrap_host.private_ip
+
+    logging.info('Preparing bootstrap host for upgrade')
+    with open(cluster.ssher.key_path, 'r') as key_fh:
+        ssh_key = key_fh.read()
+    installer = test_util.installer_api_test.DcosCliInstaller()
+    with cluster.ssher.tunnel(cluster.bootstrap_host) as bootstrap_host_tunnel:
+        installer.setup_remote(
+            tunnel=bootstrap_host_tunnel,
+            installer_path=cluster.ssher.home_dir + '/dcos_generate_config.sh',
+            download_url=installer_url,
+        )
+        installer.genconf(
+            bootstrap_url=bootstrap_url,
+            zk_host=cluster.bootstrap_host.private_ip + ':2181',
+            master_list=[h.private_ip for h in cluster.masters],
+            agent_list=[h.private_ip for h in cluster.agents],
+            public_agent_list=[h.private_ip for h in cluster.public_agents],
+            ip_detect='aws',
+            ssh_user=cluster.ssher.user,
+            ssh_key=ssh_key,
+            add_config_path=add_config_path,
+            rexray_config_preset='aws',
+        )
+        # Remove docker (and associated journald) restart from the install
+        # script. This prevents Docker-containerized tasks from being killed
+        # during agent upgrades.
+        bootstrap_host_tunnel.remote_cmd([
+            'sudo', 'sed',
+            '-i',
+            '-e', '"s/systemctl restart systemd-journald//g"',
+            '-e', '"s/systemctl restart docker//g"',
+            cluster.ssher.home_dir + '/genconf/serve/dcos_install.sh',
+        ])
+        run_bootstrap_nginx(bootstrap_host_tunnel, cluster.ssher.home_dir)
+
+    upgrade_ordering = [
+        ('master', 'master', master_upgrade_order(cluster)),
+        ('slave', 'agent', cluster.agents),
+        ('slave_public', 'public agent', cluster.public_agents),
+    ]
+    for role, role_name, hosts in upgrade_ordering:
+        logging.info('Upgrading {} nodes: {}'.format(role_name, repr(hosts)))
+        for host in hosts:
+            logging.info('Upgrading {}: {}'.format(role_name, repr(host)))
+            with cluster.ssher.tunnel(host) as tunnel:
+                upgrade_host(tunnel, role, bootstrap_url)
+
+            if role == 'master':
+                wait_metric = 'registrar/log/recovered'
+            else:
```

Comment:
nit: 
```
wait_metric = {
    'master': 'registrar/log/recovered',
    'agent': 'slave/registered',
    'agent_public': 'slave/registered'
}[role]
```

I really prefer to be explicit about the expected set of cases. The re-formating I give is what is commonly used in enterprise for this sort of thing. If one is commonly used and wants a shorthand can make a local variable. Really don't want new cases to be added and silently fall through the agent code path here leading to a lot of painful debugging.